File: app/src/main/java/com/davismiyashiro/expenses/view/opentab/ParticipantFragment.java
package com.davismiyashiro.expenses.view.opentab;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.davismiyashiro.expenses.R;
import com.davismiyashiro.expenses.datatypes.Participant;
import com.davismiyashiro.expenses.datatypes.Tab;
import com.davismiyashiro.expenses.injection.App;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import timber.log.Timber;

/**
 * A fragment representing a list of Items.
 * <p/>
 * Activities containing this fragment MUST implement the {@link OnParticipantListFragmentInteractionListener}
 * interface.
 */
public class ParticipantFragment extends Fragment implements ParticipantInterfaces.ParticipantView {

    private static final String TAB_PARAM = "com.davismiyashiro.expenses.view.opentab.ParticipantFragment";
    private OnParticipantListFragmentInteractionListener mListener;

    @Inject
    ParticipantInterfaces.UserActionsListener mPresenter;

    private Tab mTab;
    private RecyclerView recyclerView;
    private ParticipantRecyclerViewAdapter mRecyclerAdapter;

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public ParticipantFragment() {
    }

    public static ParticipantFragment newInstance(Tab tab) {
        ParticipantFragment fragment = new ParticipantFragment();
        Bundle args = new Bundle();
        args.putParcelable(TAB_PARAM, tab);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        Timber.d("onAttach");
        if (context instanceof OnParticipantListFragmentInteractionListener) {
            mListener = (OnParticipantListFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnParticipantListFragmentInteractionListener");
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Timber.d("onCreate");

        if (getArguments() != null) {
            mTab = getArguments().getParcelable(TAB_PARAM);
        }

        mRecyclerAdapter = new ParticipantRecyclerViewAdapter(new ArrayList<Participant>(), mListener, getActivity());
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        ((App)getActivity().getApplication()).getComponent().inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        Timber.d("onCreateView");

        View view = inflater.inflate(R.layout.fragment_participant_list, container, false);

        recyclerView = (RecyclerView) view.findViewById(R.id.list);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        RecyclerView.ItemDecoration itemDecoration = new
                DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL);
        recyclerView.addItemDecoration(itemDecoration);
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setAdapter(mRecyclerAdapter);
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        Timber.d("onResume");

        mPresenter.loadParticipants(mTab);
    }

    @Override
    public void onStart() {
        super.onStart();
        mPresenter.setParticipantView(this);
    }

    @Override
    public void showParticipants(List<Participant> participants) {
        mRecyclerAdapter.replaceData(participants);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        Timber.d("onDetach");
        mListener = null;
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p/>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnParticipantListFragmentInteractionListener {
        void onParticipantListFragmentInteraction(Participant item);

        void onParticipantListFragmentLongClick(Participant item);
    }
}

File: app/src/main/java/com/davismiyashiro/expenses/view/opentab/ParticipantFragment.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Davis Miyashiro
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.davismiyashiro.expenses.view.opentab

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.davismiyashiro.expenses.R
import com.davismiyashiro.expenses.datatypes.Participant
import com.davismiyashiro.expenses.datatypes.Tab
import com.davismiyashiro.expenses.injection.App
import com.davismiyashiro.expenses.view.addparticipant.ParticipantActivity

import java.util.ArrayList

import javax.inject.Inject

import timber.log.Timber

/**
 * A fragment representing a list of Items.
 *
 *
 * Activities containing this fragment MUST implement the [OnParticipantListFragmentInteractionListener?]
 * interface.
 */
/**
 * Mandatory empty constructor for the fragment manager to instantiate the
 * fragment (e.g. upon screen orientation changes).
 */
class ParticipantFragment : Fragment(),
        ParticipantInterfaces.ParticipantView,
        ParticipantRecyclerViewAdapter.OnParticipantListFragmentInteractionListener {

    @Inject
    internal lateinit var mPresenter: ParticipantInterfaces.UserActionsListener

    private lateinit var mTab: Tab
    private lateinit var recyclerView: RecyclerView
    private lateinit var mRecyclerAdapter: ParticipantRecyclerViewAdapter

    override fun onAttach(context: Context) {
        super.onAttach(context)
        Timber.d("onAttach")
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Timber.d("onCreate")

        if (arguments != null && arguments is Bundle) {
            mTab = (arguments as Bundle).getParcelable(TAB_PARAM)
        }

        mRecyclerAdapter = ParticipantRecyclerViewAdapter(ArrayList(), this, activity as Activity)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        (activity?.application as App).component.inject(this)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        Timber.d("onCreateView")

        val view = inflater.inflate(R.layout.fragment_participant_list, container, false)

        recyclerView = view.findViewById<View>(R.id.list) as RecyclerView
        recyclerView.layoutManager = LinearLayoutManager(context)

        val itemDecoration = DividerItemDecoration(context, DividerItemDecoration.VERTICAL)
        recyclerView.addItemDecoration(itemDecoration)
        recyclerView.itemAnimator = DefaultItemAnimator()
        recyclerView.adapter = mRecyclerAdapter
        return view
    }

    override fun onResume() {
        super.onResume()
        Timber.d("onResume")

        mPresenter.loadParticipants(mTab)
    }

    override fun onStart() {
        super.onStart()
        mPresenter.setParticipantView(this)
    }

    override fun showParticipants(participants: MutableList<Participant>) {
        mRecyclerAdapter.replaceData(participants)
    }

    override fun onDetach() {
        super.onDetach()
        Timber.d("onDetach")
    }

    override fun onParticipantListFragmentInteraction(item: Participant) {
        activity?.applicationContext?.let {
            startActivity(ParticipantActivity.newInstance(it , mTab, item))
        }
    }

    override fun onParticipantListFragmentLongClick(item: Participant) {
        mPresenter.removeParticipant(item)
    }

    companion object {

        private const val TAB_PARAM = "com.davismiyashiro.expenses.view.opentab.ParticipantFragment"

        fun newInstance(tab: Tab): ParticipantFragment {
            val fragment = ParticipantFragment()
            val args = Bundle()
            args.putParcelable(TAB_PARAM, tab)
            fragment.arguments = args
            return fragment
        }
    }
}
